### PR TITLE
accel/tcg: remove implied BQL from cpu_handle_interrupt/exception path

### PR DESCRIPTION
**Pull request by rf972**
The purpose of this change is to set the groundwork
so that an arch could move towards removing
the BQL from the cpu_handle_interrupt/exception paths.

The BQL is a bottleneck in scaling to more cores.
And this cpu_handle_interrupt/exception path is one of
the key BQL users as measured by the QEMU sync profiling (qsp).

We have chosen to break up the process of removing
BQL from this path into two pieces:

1) Changes to the core/common functions of cpu_handle_interrupt/exception 
   to drop the holding of the BQL.  The holding of the BQL is pushed down 
   to the per-arch implementation code.
   This set of changes is handled in this patch.
   
   This approach of pushing the BQL down to the per arch functions was 
   suggested by Paolo Bonzini.
   For reference, here are two key posts in the discussion, explaining
   the reasoning/benefits of this approach.
   https://lists.gnu.org/archive/html/qemu-devel/2020-07/msg08731.html
   https://lists.gnu.org/archive/html/qemu-devel/2020-08/msg00044.html

2) Removing the BQL from the per-arch functions.
  Since the arch now has the code that grabs the BQL, each arch can 
  change its use of the BQL for interrupts independently.   
  We leave it up to the arch to make the change at the time that makes sense.

It is worth mentioning that we are working on per-arch changes
in line with 2), and plan to submit these.
In other words, we plan to set the groundwork with this
patch series and then will take advantage of it in later series.

### CODE REVIEW

**Review by PeterPuhov-Linaro — APPROVED** [review]

**Review by PeterPuhov-Linaro — APPROVED** [review]

**Review by PeterPuhov-Linaro — APPROVED** [review]

**Comment by PeterPuhov-Linaro** [comment]
LGTM, the only thing they may not like is:
if (bql) {
        qemu_mutex_unlock_iothread();
    }
instead of 
if (bql)
        qemu_mutex_unlock_iothread();

Linux kernel folks a little bit pedantic about it. I do not know what QEMU folks think about it :)
